### PR TITLE
207 implement lp rewards on amm

### PR DESCRIPTION
**Pull request by RustNinja**
PR related to https://github.com/pendulum-chain/pendulum/issues/207

The decision to use bifrost staking pallet looks attractive because [Zenlink UI](https://dex.zenlink.pro/#/earn/stake) will support by default this type of implementation for stake(earn) rewards from LP tokens

Foucoco runtime uses forked version of bifrost `farming` pallet.  [original code link](https://github.com/bifrost-finance/bifrost/tree/develop/pallets/farming)
[Pendulum farming](https://github.com/pendulum-chain/bifrost) pallet(forked) was changed and adapted to support spacewalk `CurrencyId`.

- add farming pallet to foucoco runtime
- add farming runtime RPC endpoints for zenlink UI

Probably better to move [Pendulum farming](https://github.com/pendulum-chain/bifrost) pallet(forked) directly to pendulum chain repo as a new pallet. (as parachain-staking from KILT was added)

Ready for review and Integration with zenlink UI once they merged new version of UI to main branch.

### CODE REVIEW

File: runtime/foucoco/src/lib.rs [hunk 1]
```diff
@@ -1452,6 +1460,30 @@ impl currency::Config for Runtime {
 	type AmountCompatibility = primitives::StellarCompatibility;
 }
 
+parameter_types! {
+	pub const FarmingKeeperPalletId: PalletId = PalletId(*b"bf/fmkpr");
+	pub const FarmingRewardIssuerPalletId: PalletId = PalletId(*b"bf/fmrir");
+}
+parameter_types! {
+	pub BifrostTreasuryAccount: AccountId = TreasuryPalletId::get().into_account_truncating();
```
Comment (ebma) [inline runtime/foucoco/src/lib.rs:1468]:
```suggestion
	pub FoucocoTreasuryAccount: AccountId = TreasuryPalletId::get().into_account_truncating();
```

File: runtime/foucoco/src/lib.rs [hunk 2]
```diff
@@ -1452,6 +1460,30 @@ impl currency::Config for Runtime {
...
+}
+
+type FarmingControlOrigin = EitherOfDiverse<
+	EnsureRoot<AccountId>,
+	pallet_collective::EnsureProportionAtLeast<AccountId, CouncilCollective, 3, 5>,
```
Comment (ebma) [inline runtime/foucoco/src/lib.rs:1473]: This seems worthy of a discussion. @Rust-Ninja I assume you chose this because you think it's a good fit? Or was this defined elsewhere and you adopted it?

File: runtime/foucoco/src/lib.rs [hunk 3]
```diff
@@ -110,6 +112,9 @@ use sp_core::crypto::UncheckedFrom;
 // XCM Imports
 use xcm_executor::XcmExecutor;
 
+/// Balancer pool ID.
+pub type PoolId = u32; //pool id for farming rpc api
```
[the remark this answers is not in the document]
Comment (ebma) [reply runtime/foucoco/src/lib.rs:116]: I assume it's the datatype used for generating/assigning ID's to farming pools? If so, I think the comments are a bit misleading.

Maybe rephrase the comments to say that it's the 
```
/// Type for IDs of farming pools

**Comment by ebma** [comment]
I also checked your changes to the bifrost farming pallet [here](https://github.com/pendulum-chain/bifrost/tree/polkadot-v0.9.37) and it looks good to me @RustNinja .

> Probably better to move [Pendulum farming](https://github.com/pendulum-chain/bifrost) pallet(forked) directly to pendulum chain repo as a new pallet. (as parachain-staking from KILT was added)

But we still have to take a decision here: Should we move the pallet to this repository or keep it as a fork? It indeed seems to be necessary to fork the pallet because we need to use our own currency ID. But in theory, we can create a PR with the abstraction that @RustNinja added to the original bifrost repo, because they just need to have this extra `Config` parameter. But since it's not necessary for them I don't know if they are willing to add that change (for us).
@TorstenStueber do you have an opinion on this? Should we try to get bifrost to merge that change for us or stick with the fork?

File: runtime/foucoco/src/lib.rs (hunk 2 above)
Comment (TorstenStueber) [reply runtime/foucoco/src/lib.rs:1473]: I prefer if we just set this to root for now.

**Comment by TorstenStueber** [comment]
There is a good chance that they find this change unnecessary or that accepting the change will take a while. Therefore, I suggest to open a PR with Bifrost but in the meantime use [our fork](https://github.com/pendulum-chain/bifrost). If they won't accept it or don't find it necessary, then we should move our version of the pallet into the Pendulum repo.

**Comment by RustNinja** [comment]
> There is a good chance that they find this change unnecessary or that accepting the change will take a while. Therefore, I suggest to open a PR with Bifrost but in the meantime use [our fork](https://github.com/pendulum-chain/bifrost). If they won't accept it or don't find it necessary, then we should move our version of the pallet into the Pendulum repo.

@TorstenStueber 
Should I open pr to `main` bifrost branch with polkadot 0.9.38 and `polkadot 0.9.37` also? 
Because for now we need a fix in `polkadot 0.9.37` bifrost branch also.

**Comment by ebma** [comment]
Ah right, I forgot that we also have to change the dependency versions for ORML-related crates.
But we could actually also go the easy route and just add some `patch` statements to our root `Cargo.toml` which would make sure that every dependency is using the same version of OMRL, even transitive/peer-dependencies. We would run into problems if some breaking changes were introduced between the versions though.

@RustNinja about the PR question, maybe we should create the PR to the `main` branch and ask them to also merge the change into `polkadot-v0.9.37` and `polkadot-v0.9.38`?

**Comment by ebma** [comment]
> Then I first will change orml dependency in spacewalk and pendulum repo and create PR to use crates version instead of git.

Do you mean you create a PR with the `patch` statements? Or what exactly do you want to change?

**Comment by TorstenStueber** [comment]
@RustNinja 
> Should I open pr to main bifrost branch with polkadot 0.9.38 and polkadot 0.9.37 also?

I think, 0.9.38 is enough – we would need to update soon anyway.

@RustNinja @ebma 
> But we could actually also go the easy route and just add some patch statements to our root Cargo.toml

I think that makes a lot of sense. This is why I would like to avoid to use too many third party pallets – getting all dependencies straight (just the Polkadot version alone) is quite some effort.

File: runtime/foucoco/src/lib.rs (hunk 2 above)
Comment (RustNinja) [reply runtime/foucoco/src/lib.rs]: Done.

File: runtime/foucoco/src/lib.rs [hunk 4]
```diff
@@ -1452,6 +1460,30 @@ impl currency::Config for Runtime {
 	type AmountCompatibility = primitives::StellarCompatibility;
 }
 
+parameter_types! {
+	pub const FarmingKeeperPalletId: PalletId = PalletId(*b"bf/fmkpr");
+	pub const FarmingRewardIssuerPalletId: PalletId = PalletId(*b"bf/fmrir");
+}
+parameter_types! {
+	pub BifrostTreasuryAccount: AccountId = TreasuryPalletId::get().into_account_truncating();
+}
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply runtime/foucoco/src/lib.rs]: Done.

File: runtime/foucoco/src/lib.rs [hunk 5]
```diff
@@ -1437,6 +1443,8 @@ impl currency::CurrencyConversion<currency::Amount<Runtime>, CurrencyId> for Cur
 	}
 }
 
+parameter_types! {}
+
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply runtime/foucoco/src/lib.rs]: Done.

File: runtime/foucoco/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RustNinja) [reply runtime/foucoco/src/lib.rs:116]: Done.

File: runtime/foucoco/Cargo.toml [hunk 6]
```diff
@@ -129,6 +129,9 @@ orml-xtokens = { git = "https://github.com/open-web3-stack/open-runtime-module-l
 zenlink-protocol = { git = "https://github.com/zenlinkpro/Zenlink-DEX-Module", default-features = false, branch = "polkadot-v0.9.37" }
 zenlink-protocol-runtime-api = { git = "https://github.com/zenlinkpro/Zenlink-DEX-Module", default-features = false, branch = "polkadot-v0.9.37" }
 
+bifrost-farming = { git = "https://github.com/pendulum-chain/bifrost", default-features = false, branch = "introduce-abstraction-for-currency-id" }
+bifrost-farming-rpc-runtime-api = { git = "https://github.com/pendulum-chain/bifrost", default-features = false, branch = "introduce-abstraction-for-currency-id" }
```
Comment (ebma) [inline runtime/foucoco/Cargo.toml:133]: Let's use `polkadot-v0.9.37` now that you merged the PR

**Review by ebma — CHANGES_REQUESTED** [review]

File: runtime/foucoco/src/lib.rs [hunk 7]
```diff
@@ -1461,23 +1459,17 @@ impl currency::Config for Runtime {
...
+	pub const FarmingKeeperPalletId: PalletId = PalletId(*b"fo/fmkpr");
+	pub const FarmingRewardIssuerPalletId: PalletId = PalletId(*b"fo/fmrir");
 	pub FoucocoTreasuryAccount: AccountId = TreasuryPalletId::get().into_account_truncating();
 }
-
-type FarmingControlOrigin = EitherOfDiverse<
-	EnsureRoot<AccountId>,
-	pallet_collective::EnsureProportionAtLeast<AccountId, CouncilCollective, 3, 5>,
->;
+parameter_types! {}
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply runtime/foucoco/src/lib.rs:1466]: Done.

File: node/src/rpc.rs [hunk 8]
```diff
@@ -23,9 +23,15 @@ use spacewalk_primitives::{
...
 
+use bifrost_farming_rpc_api::{FarmingRpc, FarmingRpcApiServer};
+use bifrost_farming_rpc_runtime_api::FarmingRuntimeApi;
+
 /// A type representing all RPC extensions.
 pub type RpcExtension = jsonrpsee::RpcModule<()>;
 
+/// Type for IDs of farming pools
+pub type PoolId = u32; //pool id for farming rpc api
+
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply node/src/rpc.rs]: Done.

File: node/src/rpc.rs [hunk 9]
```diff
@@ -128,3 +134,70 @@ where
 
 	Ok(module)
 }
+
+pub fn create_full_spacewalk_foucoco<C, P>(
+	deps: FullDeps<C, P>,
+) -> Result<RpcExtension, Box<dyn std::error::Error + Send + Sync>>
+where
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply node/src/rpc.rs]: Done.

File: Cargo.toml [hunk 10]
```diff
@@ -13,3 +13,15 @@ members = [
 	"runtime/development",
 	"runtime/integration-tests/pendulum",
 ]
+
+# need this because of bifrost farming dependency in runtime
+[patch."https://github.com/open-web3-stack/open-runtime-module-library"] 
+orml-traits = { git = "https://github.com/open-web3-stack//open-runtime-module-library", branch = "polkadot-v0.9.37" }
```
[the remark this answers is not in the document]
Comment (RustNinja) [reply Cargo.toml:20]: Done.

File: runtime/foucoco/Cargo.toml (hunk 6 above)
Comment (RustNinja) [reply runtime/foucoco/Cargo.toml]: Done.

File: runtime/common/src/lib.rs [hunk 11]
```diff
@@ -14,6 +14,9 @@ pub type Signature = MultiSignature;
 /// to the public key of our transaction signing scheme.
 pub type AccountId = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;
 
+/// Type for IDs of farming pools
+pub type PoolId = u32; //pool id for farming rpc api
```
Comment (ebma) [inline runtime/common/src/lib.rs:18]: Nit: The other comment is probably not necessary anymore.
```suggestion
pub type PoolId = u32;
```

**Review by ebma — COMMENTED** [review]
Thanks @RustNinja 👍 

Maybe we should now also rename the `start_node_impl` functions
- `start_node_impl()  ->`start_node_impl_pendulum()` 
- `start_node_impl_spacewalk_foucoco()` -> `start_node_impl_foucoco()`
- `start_node_impl_spacewalk_amplitude()` -> `start_node_impl_amplitude()`

Sorry for coming back with this again, I did not realize that we should also rename them before.

**Review by ebma — APPROVED** [review]
LGTM, let's wait for CI and @TorstenStueber's approval before merge

**Review by TorstenStueber — APPROVED** [review]
